src/channel.c

```diff
@@ -0,0 +1,232 @@
+
+#define TOO_TRICKY (0)
+#define CHANNEL_VERBOSE (0)
+
+typedef struct channel_t {
+   struct ringbuffer_t* rb;
+   pthread_mutex_t mutex;
+   pthread_cond_t read_avail_cond;
+#if TOO_TRICKY
+   pthread_cond_t write_avail_cond;
+#endif
+   int closed;
+   int drained;
+   uint32_t num_items;
+   int refcount;
+   size_t size_increment;
+} channel_t;
+
+static void channel_init_queue(channel_t *channel, size_t size) {
+   // init queue mutex
+   pthread_mutexattr_t mutex_attr;
+   pthread_mutexattr_init(&mutex_attr);
+   pthread_mutexattr_settype(&mutex_attr, PTHREAD_MUTEX_RECURSIVE);
+   pthread_mutex_init(&channel->mutex, &mutex_attr);
+
+   // init condition variables
+   pthread_cond_init(&channel->read_avail_cond, NULL);
+#if TOO_TRICKY
+   pthread_cond_init(&channel->write_avail_cond, NULL);
+#endif
+
+   // init ring buffer
+   channel->rb = ringbuffer_create(size);
+}
+
+int channel_create(lua_State *L) {
+   channel_t *channel = calloc(1, sizeof(channel_t));
+   channel->refcount = 1;
+   channel->closed = 0;
+   channel->drained = 0;
+   channel_t **ud = (channel_t **)lua_newuserdata(L, sizeof(channel_t*));
+   channel_init_queue(channel, DEFAULT_CHANNEL_SIZE);
+   channel->size_increment = DEFAULT_CHANNEL_SIZE;
+   *ud = channel;
+   luaL_getmetatable(L, "ipc.channel");
+   lua_setmetatable(L, -2);
+   return 1;
+}
+
+int channel_close(lua_State *L) {
+   channel_t *channel = *(channel_t **)lua_touserdata(L, 1);
+   if (!channel) return LUA_HANDLE_ERROR_STR(L, "invalid channel");
+   pthread_mutex_lock(&channel->mutex);
+   if (!channel->closed) {
+      channel->closed = 1;
+      if (channel->num_items == 0) {
+         channel->drained = 1;
+      }
+      pthread_cond_broadcast(&channel->read_avail_cond);
+   }
+   pthread_mutex_unlock(&channel->mutex);
+   return 0;
+}
+
+int channel_closed(lua_State *L) {
+   channel_t *channel = *(channel_t **)lua_touserdata(L, 1);
+   if (!channel) return LUA_HANDLE_ERROR_STR(L, "invalid channel");
+   pthread_mutex_lock(&channel->mutex);
+   lua_pushboolean(L, channel->closed);
+   pthread_mutex_unlock(&channel->mutex);
+   return 1;
+}
+
+int channel_drained(lua_State *L) {
+   channel_t *channel = *(channel_t **)lua_touserdata(L, 1);
+   if (!channel) return LUA_HANDLE_ERROR_STR(L, "invalid channel");
+   pthread_mutex_lock(&channel->mutex);
+   lua_pushboolean(L, channel->drained);
+   pthread_mutex_unlock(&channel->mutex);
+   return 1;
+}
+
+int channel_read(lua_State *L) {
+   channel_t *channel = *(channel_t **)lua_touserdata(L, 1);
+   if (!channel) return LUA_HANDLE_ERROR_STR(L, "invalid channel");
+   int doNotBlock = luaT_optboolean(L, 2, 0);
+   pthread_mutex_lock(&channel->mutex);
+   while (1) {
+      if (channel->num_items) {
+         if (channel->closed && channel->num_items == 1) {
+            channel->drained = 1;
+            pthread_cond_broadcast(&channel->read_avail_cond);
+         }
+         if (channel->closed) {
+            lua_pushstring(L, ":closed");
```

Comment:
See the comment below about how to manage these strings in the user code, but here I would suggest you do something like:

#define STATUS_CLOSED ":closed"